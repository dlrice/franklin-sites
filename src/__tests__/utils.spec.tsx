/* eslint-disable no-multi-spaces */
import { render } from '@testing-library/react';

import {
  getFlattenedPaths,
  getLastIndexOfSubstringIgnoreCase,
  restructureFlattenedTreeDataForAutocomplete,
  restructureFlattenedTreeItemsForAutocomplete,
  highlightSubstring,
  capitaliseFirstLetter,
  tidyUrlString,
} from '../utils';

import { treeData } from '../mock-data/tree-data';

test('should get all paths', () => {
  const path = getFlattenedPaths(treeData);
  expect(path).toEqual([
    [
      {
        label: 'Item 1',
        id: 'item_1',
      },
      {
        label: 'Item 1a',
        id: 'item_1a',
      },
    ],
    [
      {
        label: 'Item 1',
        id: 'item_1',
      },
      {
        label: 'Item 1b',
        id: 'item_1b',
      },
      {
        label: 'Item 1b A',
        id: 'item_1b_A',
      },
    ],
    [
      {
        label: 'Item 1',
        id: 'item_1',
      },
      {
        label: 'Item 1b',
        id: 'item_1b',
      },
      {
        label: 'Item 1b B',
        id: 'item_1b_B',
      },
    ],
    [
      {
        label: 'Some Item 2',
        id: 'item_2',
      },
    ],
  ]);
});

test('should find the correct path', () => {
  const path = getFlattenedPaths(treeData, 'item_1b_B');
  expect(path).toEqual([
    [
      {
        label: 'Item 1',
        id: 'item_1',
      },
      {
        label: 'Item 1b',
        id: 'item_1b',
      },
      {
        label: 'Item 1b B',
        id: 'item_1b_B',
      },
    ],
  ]);
});

test('should not find any paths', () => {
  const path = getFlattenedPaths(treeData, 'The unfindable');
  expect(path).toEqual([]);
});

test('should find the index of the last substring', () => {
  const index = getLastIndexOfSubstringIgnoreCase('StringSTRING', 'String');
  expect(index).toEqual(6);
});

test('should not find the index of the last string', () => {
  const index = getLastIndexOfSubstringIgnoreCase('StringSTRING', 'unfindable');
  expect(index).toEqual(-1);
});

test('should prepare all tree data for autocomplete', () => {
  const flatPaths = getFlattenedPaths(treeData);
  const data = restructureFlattenedTreeDataForAutocomplete(flatPaths);
  expect(data).toEqual([
    {
      id: 'item_1a',
      pathLabel: 'Item 1 / Item 1a',
      itemLabel: 'Item 1a',
    },
    {
      id: 'item_1b_A',
      pathLabel: 'Item 1 / Item 1b / Item 1b A',
      itemLabel: 'Item 1b A',
    },
    {
      id: 'item_1b_B',
      pathLabel: 'Item 1 / Item 1b / Item 1b B',
      itemLabel: 'Item 1b B',
    },
    {
      id: 'item_2',
      pathLabel: 'Some Item 2',
      itemLabel: 'Some Item 2',
    },
  ]);
});

test('should prepare flattened tree data items for autocomplete', () => {
  const items = [
    {
      label: 'Item 1',
      id: 'item_1',
    },
    {
      label: 'Item 1a',
      id: 'item_1a',
    },
  ];
  const data = restructureFlattenedTreeItemsForAutocomplete(items);
  expect(data).toEqual({
    id: 'item_1a',
    pathLabel: 'Item 1 / Item 1a',
    itemLabel: 'Item 1a',
  });
});

describe('capitaliseFirstLetter', () => {
  expect(capitaliseFirstLetter(null)).toBe(null);
  expect(capitaliseFirstLetter('')).toBe('');
  expect(capitaliseFirstLetter('a')).toBe('A');
  expect(capitaliseFirstLetter('A')).toBe('A');
  expect(capitaliseFirstLetter('hello')).toBe('Hello');
  expect(capitaliseFirstLetter('Hello')).toBe('Hello');
});

describe('highlightSubstring', () => {
  const string = 'Item 1 / Item 1a';
  test('should highlight substring (case insensitive', () => {
    const { asFragment } = render(
      highlightSubstring(string, 'item 1') as JSX.Element
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('should return string if no substring found', () => {
    expect(highlightSubstring(string, 'zap')).toEqual(string);
  });
});

describe('tidyUrlString', () => {
  it('should remove http:// and any trailing slashes', () => {
    expect(tidyUrlString('http://www.ebi.ac.uk//')).toEqual('www.ebi.ac.uk');
  });
  it('should remove https', () => {
    expect(tidyUrlString('https://ebi.ac.uk')).toEqual('ebi.ac.uk');
  });
  it('should retain www', () => {
    expect(tidyUrlString('www.ebi.ac.uk')).toEqual('www.ebi.ac.uk');
  });
  it('should leave forward slashes in the middle of the url', () => {
    expect(tidyUrlString('https://www.ebi.ac.uk/uniprot/TrEMBLstats')).toEqual(
      'www.ebi.ac.uk/uniprot/TrEMBLstats'
    );
  });
  it('should return the same url if it is already tidy', () => {
    expect(tidyUrlString('ebi.ac.uk')).toEqual('ebi.ac.uk');
  });
  it('should handle a schema-relative URL', () => {
    expect(tidyUrlString('//www.ebi.ac.uk/')).toEqual('www.ebi.ac.uk');
  });
});
