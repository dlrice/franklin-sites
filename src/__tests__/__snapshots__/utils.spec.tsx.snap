// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`highlightSubstring should highlight substring (case insensitive 1`] = `
<DocumentFragment>
  Item 1 / 
  <b>
    Item 1
  </b>
  a
</DocumentFragment>
`;
